### PR TITLE
Dao

### PR DESCRIPTION
**Pull request by slammy-dev**
DAO classes, changed db.props and pom, added abstract class for testing

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -115,6 +115,20 @@
...
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>1.4.196</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-test</artifactId>
+            <version>LATEST</version>
```
Comment (SMishAlex) [inline pom.xml]: it's better to use same version as spring.ver

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        try {
+            PreparedStatement ps = connection.prepareStatement(sql);
+            ResultSet rs = ps.executeQuery();
+            while (rs.next()) {
+                Course course = parseCourse(rs);
+                if (course != null) courses.add(course);
+            }
+
+        } catch (SQLException e) {
+            log.error(e);
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: throw some exception here with `e`  as cause

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return courses;
+    }
+
+    @Override
+    public void updateById(int id, Course entry) {
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: why don't you get id from the entry?

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            connection.prepareStatement(sql).execute();
+        } catch (SQLException e) {
+            log.error(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+    }
+
+    @Override
+    public void insert(Course entry) {
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: create and update methods usually returns updated (created) values

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    @Autowired
+    public StudentCourseDAO(ConnectionPool connectionPool) {
+        this.connectionPool = connectionPool;
+    }
+
+    @Override
+    public List<StudentCourse> getAllEntries() {
+        Connection connection = connectionPool.getConnection();
+        String sql = String.format("SELECT id, courseId, studentId, studentMark, studentFeedback FROM %s.student_course", databaseName);
+        List<StudentCourse> scList = new ArrayList<>();
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: funny naming

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+@Log4j2
+@Component
+public class CourseDAO implements DAO<Course> {
+
+    @Value("${db.table}")
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: rename to db.name

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    @Autowired
+    public CourseDAO(ConnectionPool connectionPool) {
+        this.connectionPool = connectionPool;
+    }
+
+    @Override
+    public List<Course> getAllEntries() {
+        Connection connection = connectionPool.getConnection();
+        String sql = String.format("SELECT courseId, name, tutorId, annotation, status FROM %s.courses", databaseName);
+        List<Course> courses = new ArrayList<>();
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: are we going to have duplicates here? if not, use Set then

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java (hunk 3 above)
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: should return updated Course

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        } catch (SQLException e) {
+            log.error(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+    }
+
+    @Override
+    public void removeById(int id) {
+        Connection connection = connectionPool.getConnection();
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: extract it as a class field and the reuse this variable in all the methods

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                String annotation = rs.getString("annotation");
+                CourseStatus status;
+                switch (rs.getString("status").toLowerCase()) {
+                    case "active":
+                        status = CourseStatus.ACTIVE;
+                        break;
+                    case "finished":
+                        status = CourseStatus.FINISHED;
+                        break;
+                    case "planned":
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: what are we doing if planned? and there should be 'break' as well

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+@Log4j2
+@Component
+public class StudentCourseDAO implements DAO<StudentCourse> {
+
+    @Value("${db.table}")
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: same about rename, it's not clear what exactly you are going to use. table is a part of DB

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    private String databaseName;
+
+    private ConnectionPool connectionPool;
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: make it final

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+    @Autowired
+    public StudentCourseDAO(ConnectionPool connectionPool) {
+        this.connectionPool = connectionPool;
+    }
+
+    @Override
+    public List<StudentCourse> getAllEntries() {
+        Connection connection = connectionPool.getConnection();
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: extract as class field and reuse

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            log.error(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return scList;
+    }
+
+    public List<Course> getAllCoursesByStudent(int studentId) {
+        Connection connection = connectionPool.getConnection();
+        List<Course> courseList = new ArrayList<>();
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: Maybe Set?

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+            }
+        } catch (SQLException e) {
+            log.error(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return scList;
+    }
+
+    public List<Course> getAllCoursesByStudent(int studentId) {
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: according to the name of the method you should pass Student as the argument

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        String sql = String.format("SELECT courseId from %s.student_course where studentId=%d", databaseName, studentId);
+        try {
+            ResultSet rs = connection.prepareStatement(sql).executeQuery();
+            while (rs.next()) {
+                int id = rs.getInt("courseId");
+                sql = String.format("SELECT courseId, name, tutorId, annotation, status FROM %s.courses where courseId=%d",
+                        databaseName, id);
+                ResultSet userResult = connection.prepareStatement(sql).executeQuery();
+                courseList.add(CourseDAO.parseCourse(userResult));
+            }
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: you have to rewrite everything above using JOIN, it's totally wrong!

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+            log.error(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return courseList;
+    }
+
+    public List<User> getAllStudentsByCourse(int courseId) {
+        Connection connection = connectionPool.getConnection();
+        List<User> studentList = new ArrayList<>();
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: maybe Set?

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+        String sql = String.format("SELECT studentId from %s.student_course where courseId=%d", databaseName, courseId);
+        try {
+            ResultSet rs = connection.prepareStatement(sql).executeQuery();
+            while (rs.next()) {
+                int id = rs.getInt("studentId");
+                sql = String.format("SELECT userId, email, pass, name, lastName, role FROM %s.users WHERE userId=%d",
+                        databaseName, id);
+                ResultSet userResult = connection.prepareStatement(sql).executeQuery();
+                studentList.add(UserDAO.parseUser(userResult));
+            }
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: same about join

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+        } catch (SQLException e) {
+            log.error(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return studentList;
+    }
+
+    @Override
+    public void updateById(int id, StudentCourse entry) {
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: return updated StudentCourse domain

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java (hunk 8 above)
Comment (SMishAlex) [reply src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: I think that Idea is take connection every time and release it then, i gues it should be here

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java (hunk 9 above)
Comment (slammy-dev) [reply src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: if planned we set CourseStatus to "planned". Else, if this field is corrupted ("plannnnned") we'll set "planned" as well by default

File: pom.xml [hunk 19]
```diff
@@ -140,7 +140,7 @@
                 <version>2.2</version>
                 <configuration>
                     <server>tomcat-development-server</server>
-                    <port>8080</port>
+                    <port>8088</port>
```
Comment (SMishAlex) [inline pom.xml]: ?

File: src/main/java/com/epam/webelecty/models/User.java [hunk 20]
```diff
@@ -16,4 +18,14 @@
     private String name;
     private String lastName;
     private UserRole role;
+
+    private String confirmPassword;
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/models/User.java]: should be changed in next version

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 21]
```diff
@@ -30,113 +29,91 @@ public CourseDAO(ConnectionPool connectionPool) {
...
                 databaseName, entry.getCourseName(), entry.getTutorId(), entry.getAnnotation(), entry.getStatus().name());
-        try {
-            connection.prepareStatement(sql).executeUpdate();
-        } catch (SQLException e) {
-            log.error(e);
-        } finally {
-            connectionPool.releaseConnection(connection);
-        }
+        executeSqlStatement(connectionPool, sql);
+        return entry;
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: it's not correct... u should read new entity from db to get it's ID

File: src/main/java/com/epam/webelecty/services/UserServiceImplementation.java [hunk 22]
```diff
@@ -0,0 +1,29 @@
+package com.epam.webelecty.services;
+
+import com.epam.webelecty.models.User;
+import com.epam.webelecty.persistence.dao.UserDAO;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserServiceImplementation  implements UserService{
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/services/UserServiceImplementation.java:10]: use auto-format

**Review by SMishAlex — COMMENTED** [review]
update and insert still not correct

File: src/main/java/com/epam/webelecty/controllers/TestController.java [hunk 23]
```diff
@@ -1,20 +1,61 @@
...
+import org.springframework.web.bind.annotation.ResponseBody;
 import org.springframework.web.servlet.ModelAndView;
 
 @Controller
 public class TestController {
 
-    @RequestMapping(value = {"/", "/index"}, method = { RequestMethod.GET })
-    public ModelAndView getWelcomePage(){
+    @Autowired
+    UserService userService;
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/controllers/TestController.java:22]: filed injection is not recommended

File: src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+        } catch (SQLException e) {
+            log.error(e);
+            throw new NoCourseFoundException(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return course;
+    }
+
+    public static Course parseCourse(ResultSet rs) {
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/CourseDAO.java]: are we using this method outside of this class? if not make it private then

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+        Connection connection = connectionPool.getConnection();
+        Set<User> studentSet = new HashSet<>();
+        String sql = String.format("SELECT users.userId, email, pass, name, lastName, role from %s.student_course JOIN " +
+                        "%s.users ON %s.student_course.studentId = %s.users.userId where courseId=%d",
+                databaseName, databaseName, databaseName, databaseName, course.getCourseId());
+        fillUserSet(connection, studentSet, sql);
+        return studentSet;
+    }
+
+    public Set<User> getAllStudentsByCourseId(int id) {
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: useless method. Above there is one that takes course as an argument and process it, why do we need another one? if we have a course already then we have it's id as well

File: src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+        } catch (SQLException e) {
+            log.error(e);
+            throw new NoStudentCourseFoundException(e);
+        } finally {
+            connectionPool.releaseConnection(connection);
+        }
+        return sc;
+    }
+
+    public static StudentCourse parseStudentCourse(ResultSet rs) {
```
Comment (petriankins) [inline src/main/java/com/epam/webelecty/persistence/dao/StudentCourseDAO.java]: is there need to make it public?

**Pull request merged**